File: sources/entity/tests/test_config.php
<?php
/**
 * Test that config.php passes through all required settings keys.
 * This prevents the bug where new settings are added to settings.json
 * but not to config.php, causing features to silently fail in production.
 *
 * Run: php tests/test_config.php
 */

$passed = 0;
$failed = 0;

function test(string $label, $expected, $actual): void
{
    global $passed, $failed;
    if ($expected === $actual) {
        echo "  PASS: {$label}\n";
        $passed++;
    } else {
        echo "  FAIL: {$label}\n    Expected: " . json_encode($expected) . "\n    Got:      " . json_encode($actual) . "\n";
        $failed++;
    }
}

echo "=== Config Key Passthrough ===\n\n";

$settings = json_decode(file_get_contents(__DIR__ . '/../php/settings.json'), true);
$config = require __DIR__ . '/../php/config.php';

// These keys MUST be present in config.php output
$requiredKeys = [
    'anthropic_api_key',
    'openai_api_key',
    'browserbase_api_key',
    'browserbase_project_id',
    'model',
    'sec_user_agent',
    'max_page_chars',
    'max_entity_names',
    'max_ciks',
    'max_ownership_levels',
    'brightdata_api_key',
    'brightdata_zone',
    'companies_house_api_key',
    'northdata_email',
    'northdata_password',
    'blocked_entity_names',
];

echo "1. All required keys present in config output:\n";
foreach ($requiredKeys as $key) {
    test("Key '{$key}' exists in config", true, array_key_exists($key, $config));
}

echo "\n2. Config values match settings.json:\n";
foreach ($requiredKeys as $key) {
    if (!isset($settings[$key])) continue;
    $settingsVal = $settings[$key];
    $configVal = $config[$key] ?? null;
    // For numeric values, settings.json has strings but config.php casts to int
    if (is_int($configVal) && is_string($settingsVal)) {
        $settingsVal = (int) $settingsVal;
    }
    test("Key '{$key}' value matches", $settingsVal, $configVal);
}

echo "\n3. Model dispatch:\n";
require_once __DIR__ . '/../php/lookup.php';
// Test with Claude model
$claudeConfig = $config;
$claudeConfig['model'] = 'claude-sonnet-4-6';
$lookup = new EntityLookup($claudeConfig);
$ref = new ReflectionClass($lookup);
$method = $ref->getMethod('isOpenAIModel');
test('claude-sonnet-4-6 is NOT OpenAI', false, $method->invoke($lookup));

// Test with OpenAI model
$openaiConfig = $config;
$openaiConfig['model'] = 'gpt-4o';
$lookup2 = new EntityLookup($openaiConfig);
test('gpt-4o IS OpenAI', true, $method->invoke($lookup2));

$openaiConfig['model'] = 'o3';
$lookup3 = new EntityLookup($openaiConfig);
test('o3 IS OpenAI', true, $method->invoke($lookup3));

$openaiConfig['model'] = 'o4-mini';
$lookup4 = new EntityLookup($openaiConfig);
test('o4-mini IS OpenAI', true, $method->invoke($lookup4));

echo "\n=== Results: {$passed} passed, {$failed} failed ===\n";
exit($failed > 0 ? 1 : 0);

File: sources/entity/app/tests/test_config.py
"""
Faithful Python port of php/tests/test_config.php.

Verifies that config.load_config() passes through all required settings keys
(Section 1 + 2), and that the model-dispatch predicate (_is_openai_model, the
Python equivalent of EntityLookup::isOpenAIModel) classifies models correctly
(Section 3). Pure/offline — no network.

conft.py puts the port dir on sys.path.
"""
import json
import os

import pytest

import config as config_mod
from agent import EntityLookup

# settings.json lives next to config.py in the port dir
_SETTINGS_PATH = os.path.join(os.path.dirname(os.path.dirname(__file__)), "settings.json")

with open(_SETTINGS_PATH) as _f:
    SETTINGS = json.load(_f)

CONFIG = config_mod.load_config()

# Exact same required-keys list as test_config.php.
REQUIRED_KEYS = [
    "anthropic_api_key",
    "openai_api_key",
    "browserbase_api_key",
    "browserbase_project_id",
    "model",
    "sec_user_agent",
    "max_page_chars",
    "max_entity_names",
    "max_ciks",
    "max_ownership_levels",
    "brightdata_api_key",
    "brightdata_zone",
    "companies_house_api_key",
    "northdata_email",
    "northdata_password",
    "blocked_entity_names",
]


# ── Section 1: all required keys present in config output ────────────────────
@pytest.mark.parametrize("key", REQUIRED_KEYS)
def test_required_key_present(key):
    assert key in CONFIG, f"Key '{key}' must exist in config output"


# ── Section 2: config values match settings.json ────────────────────────────
# Mirror the PHP: for numeric values PHP settings.json holds strings but
# config.php casts to int, so int-cast the settings value when config is int.
def _settings_val_for(key):
    settings_val = SETTINGS[key]
    config_val = CONFIG.get(key)
    if isinstance(config_val, int) and not isinstance(config_val, bool) and isinstance(settings_val, str):
        settings_val = int(settings_val)
    return settings_val


@pytest.mark.parametrize("key", [k for k in REQUIRED_KEYS if k in SETTINGS])
def test_config_value_matches_settings(key):
    assert CONFIG.get(key) == _settings_val_for(key), (
        f"Key '{key}' value must match settings.json"
    )


# ── Section 3: model dispatch (isOpenAIModel) ───────────────────────────────
def _lookup_with_model(model):
    cfg = dict(CONFIG)
    cfg["model"] = model
    return EntityLookup(cfg)


def test_claude_model_is_not_openai():
    assert _lookup_with_model("claude-sonnet-4-6")._is_openai_model() is False


def test_gpt4o_is_openai():
    assert _lookup_with_model("gpt-4o")._is_openai_model() is True


def test_o3_is_openai():
    assert _lookup_with_model("o3")._is_openai_model() is True


def test_o4_mini_is_openai():
    assert _lookup_with_model("o4-mini")._is_openai_model() is True
